### PR TITLE
Build improvements

### PR DESCRIPTION
**Pull request by peter-edge**
This PR is doing a bunch of cleanup tasks:

- Add `make shell` to drop into a shell within docker.
- Move the cache to `${HOME}/.yarpc-go`. This makes it so that when you run `make generate` or `make shell` does not overwrite the built dependencies.
- Install the docker client as part of the build. This makes it so that as long as you have a remote docker host, you can use docker even if the daemon is not installed and running, makes the docker client always be the same version, and also makes it easier to use CircleCI.
- Make the docker build more granular by installing the binaries that are not dependent on glide first.
- Have a `$(LIB)` folder that is used to download binaries, and have version-specific lib folders for each binary download.
- Split downloading a binary/tar file from installing it to `$(BIN)`.
- Do not depend on glide.lock for golang binaries if using docker, so that if you use `make generate` or `make shell`, these do not get rebuilt. glide.lock changes are already taken care of in each Dockerfile.
- Build binaries for examples before running them through service-test. The timeout we had was because `go run` was taking too long to build the binaries.
- Update service-test to take a list of `run` configurations, where the last one is waited on, and split input and output to apply to each command. This helps with a race condition we had where the server might write after the client to stdout, and makes `service-test` more generic.
- Add a debug` flag to service-test.
- Make all examples binaries follow the convention where `main` calls a function `func do() error` and `log.Fatal` on error.
- Have all examples binaries use `127.0.0.1` for addresses per Prashant’s recommendation.
- Add `deps` to `ci` dependencies except for `crossdock`.
- Have `make ci` call `make crossdock-fresh` instead of `make crossdock`. 
- Use `$(info)` instead of `echo` in the Makefile.
- Print out more information from `make help` if `SUPPRESS_DOCKER` is not set.
- Make `ERRCHECK_FLAGS` a variable that cannot be set by the user.
- Remove `SUPPRESS_COVER_PARALLEL` and always do parallel tests in scripts/cover.sh.
- Add a `V` variable for verbose logging, which right now just prints debug logs from service-test.
- Have `make examples` run in parallel.
- Split out crossdock dependencies to `build/crossdockdeps.mk` to try to help with docker caching for crossdock.
- Rename `build/test.mk` to `build/local.mk`.
- Remove old matches/files from .gitignore and .dockerignore.
- Add `${HOME}/.glide` to the Travis CI cache.
- Re-add examples to the docker build after the service-test build fix.
- Add CircleCI as an experimental build.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+          exclusive: false
+      - restore_cache:
+          key: yarpc-go
+      - run:
+          name: crossdock
+          command: make ci
+          environment:
+            - CI_TYPES: crossdock
+      - run:
+          name: 1.8 lint test exzmples
```
Comment (breerly) [inline .circleci/config.yml]: examples*

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+version: 2
+jobs:
```
Comment (breerly) [inline .circleci/config.yml]: Holy hell this file is a pleasure to read compared to .travis.yml

File: build/crossdockdeps.mk [hunk 3]
```diff
@@ -0,0 +1,20 @@
+DOCKER_COMPOSE_VERSION := 1.10.0
```
Comment (breerly) [inline build/crossdockdeps.mk]: Is this still the latest?

File: build/local.mk [hunk 4]
```diff
@@ -6,10 +6,12 @@ LINT_EXCLUDES_EXTRAS =
 GOVET_IGNORE_RULES = \
 	possible formatting directive in Error call
 
-ERRCHECK_FLAGS ?= -ignoretests
+ERRCHECK_FLAGS := -ignoretests
 ERRCHECK_EXCLUDES := \.Close\(\) \.Stop\(\)
 FILTER_ERRCHECK := grep -v $(patsubst %,-e %, $(ERRCHECK_EXCLUDES))
 
+EXAMPLES_JOBS ?= 16
```
Comment (breerly) [inline build/local.mk]: Comment here might be useful.

File: .circleci/config.yml (hunk 1 above)
Comment (peter-edge) [inline .circleci/config.yml]: Done

File: build/crossdockdeps.mk (hunk 3 above)
Comment (peter-edge) [inline build/crossdockdeps.mk]: Updated

File: build/local.mk (hunk 4 above)
Comment (peter-edge) [inline build/local.mk]: Done

**Pull request merged**